src/annotator/anchoring/trim-range.ts

```diff
@@ -0,0 +1,223 @@
+ * between `container` and `boundaryContainer`
+ */
+function trimTextContainer(
+  container: Text,
+  boundaryContainer: Text,
+  { direction = FromPosition.START, root }: TrimTextContainerOptions
+) {
+  const nodeIter = container.ownerDocument!.createNodeIterator(
+    root ?? container,
+    NodeFilter.SHOW_TEXT
+  );
+  let currentNode = nodeIter.nextNode();
+
+  while (currentNode && currentNode !== container) {
+    currentNode = nodeIter.nextNode();
+  }
+
+  if (direction === FromPosition.END) {
+    // Reverse the NodeIterator direction. This will return the same node
+    // as the previous `nextNode()` call (`container`).
+    currentNode = nodeIter.previousNode();
+  }
+
+  let trimmedOffset = -1;
+
+  const advance = () => {
+    currentNode =
+      direction === FromPosition.START
+        ? nodeIter.nextNode()
+        : nodeIter.previousNode();
+
+    if (currentNode) {
+      const nodeText = currentNode.textContent ?? '';
+      const baseOffset = direction === FromPosition.START ? 0 : nodeText.length;
+      trimmedOffset = trimTextOffset(nodeText, baseOffset, direction);
+    }
+  };
+
+  // Start the examination at the first text node before/after the `container`
+  // i.e. do not evaluate the `container` itself
+  advance();
+
+  while (
+    currentNode &&
+    trimmedOffset === -1 &&
+    currentNode !== boundaryContainer
+  ) {
+    advance();
+  }
+
+  if (currentNode && trimmedOffset >= 0) {
+    return { node: currentNode, offset: trimmedOffset };
+  } else {
+    throw new RangeError(
+      'No text nodes with non-whitespace text found in range'
+    );
+  }
+}
+
+/**
+ * Return a new DOM Range that trims `range`, ensuring that:
+ *
+ * - `startContainer` and `endContainer` Text nodes both contain at least one
+ *   non-whitespace character within the Range's text content
+ * - `startOffset` and `endOffset` both point at non-whitespace characters
+ */
+export function trimRange(range: Range): Range {
+  if (!range.toString().trim().length) {
+    throw new RangeError('Range contains no non-whitespace text');
+  }
+  if (range.startContainer.nodeType !== Node.TEXT_NODE) {
+    throw new RangeError('Range startContainer is not a text node');
+  }
+  if (range.endContainer.nodeType !== Node.TEXT_NODE) {
+    throw new RangeError('Range endContainer is not a text node');
+  }
+
+  const trimmedRange = range.cloneRange();
+
+  let startTrimmed = false;
+  let endTrimmed = false;
+
+  const nodeText = (node: Node) => {
+    return (node as Text).textContent ?? '';
+  };
+
+  const trimmedOffsets = {
+    start: trimTextOffset(
+      nodeText(range.startContainer),
+      range.startOffset,
+      FromPosition.START
+    ),
+    end: trimTextOffset(
+      nodeText(range.endContainer),
+      range.endOffset,
+      FromPosition.END
+    ),
+  };
+
+  if (trimmedOffsets.start >= 0) {
+    if (trimmedOffsets.start !== range.startOffset) {
+      trimmedRange.setStart(range.startContainer, trimmedOffsets.start);
```

Comment:
`Range.setStart` should be a no-op if the start offset is the same as the current one, so you can avoid the conditional around it.

Reply:
This is really useful info, thanks!